release: 1.4.0 (#820)
biomejs/biome: website/src/content/docs/blog/biome-wins-prettier-challenge.mdx

template: splash
url: /blog/biome-wins-prettier-challenge
description: Biome formatter now 95% compatible with Prettier
summary: Biome formatter now 95% compatible with Prettier
authorName: Emanuele Stoppa
authorAvatar: https://avatars.githubusercontent.com/u/602478?v=4
authorUrl: https://twitter.com/ematipico
coauthorName: Biome Core Team, Biome Maintainers
coauthorAvatar: /img/circle-indent-logo.svg
pubDate: 2023-11-27
layout: ../../../layouts/Blog.astro
coverImage: ../../src/assets/prettier-challenge.png
coverImageAltText: The Prettier challenge banner, with the Biome logo over it
socialImage: ../../src/assets/prettier-challenge.png
---

With the release of Biome **`v1.4.0`**, we claim the bounty of the [Prettier challenge](https://console.algora.io/challenges/prettier)!

With `v1.4.0`, you'll get a better formatter experience, more formatting options, new VSCode features, new sponsors and more!

You can upgrade Biome by running the following command:

```bash
npm install --save-dev --save-exact @biomejs/biome@1.4.0
pnpm update --save-exact @biomejs/biome@1.4.0
yarn upgrade --exact @biomejs/biome@1.4.0
```

##  Better formatter

> Biome formatter has now **over 95% in terms of compatibility** against [Prettier](https://prettier.io/)!

Reviewer: Let claim 1% more :)

```suggestion
Biome formatter has now **over 96% in terms of compatibility** against [Prettier](https://prettier.io/)!
```

Merit of challenge that was launched by [Christopher Chedeau](https://twitter.com/Vjeux/status/1722733472522142022), one of the Prettier's creators.

The challenge attracted the attention of many people, and some of them decided to contribute to Biome to claim part of the bounty. I did see something amazing: contributors had an amazing coordination, they took ownership of the tasks and delivered the solution in a matter of hours.

I believe the main factors that made this possible are three:
1. Money. It's a fact, and it's completely fine if someone decides to contribute only for earning a small stipend out of it.
2. Communication. We used [GitHub](https://github.com/biomejs/biome/issues/720) as only medium of coordination. We provided information, instructions and help on how to deliver.

> 3. Infrastructure. Biome relies on a solid testing infrastructure, built by [previous](https://github.com/MichaReiser) Rome Tools [employees](https://github.com/ematipico) and [contributors](https://github.com/IWANABETHATGUY/). It's able to catch every idempotency bug, provide granular diffs and warn the user if the emitted output is the different from the one emitted by Prettier.

Reviewer: `idempotency bug` can be confusing for newcomers. I tthink we should use `unstable fromatting bug` or `reformat bug`?

Before the challenge, Biome had roughly a compatibility rate of 85%, based on our internal metrics (JavaScript, TypeScript and JSX, on options parity). Even though 85% might seem high,
the impact of a low number such as 15% on big code bases is huge, and people might feel intimidated by so many changes, causing early adopters to receive frictions when bring Biome to their team. A member of our community shared some insights:

> > As a great example of how much even just that last 5% has improved things for large codebases (and specifically with `bracketSpacing` and now `bracketSameLine` implemented) i ran it one project in our monorepo (it's a debug build on a busy machine, that's why it took 13 seconds).

Reviewer: We could omit some details. The `13s in debug build` can provide a wrong idea of Biome perf?
```suggestion
> As a great example of how much even just that last 5% has improved things for large codebases (and specifically with `bracketSpacing` and now `bracketSameLine` implemented) i ran it one project in our monorepo [...].
```

>
> Just last week, this number `[diagnostics]` was more than 6,000. Even with the bracket options ignored, it was still more than 1000, and now there are only 200 left!

The challenge has also uncovered some cases in Prettier's emitted output that we decided to not follow. We have created a [new section](/formatter/#differences-with-prettier) in our website that explains them. Our hope is to make this section smaller with the time.

> 

Reviewer: Maybe we could also say that:

- We consider any divergence as a bug, and any tester should report divergence
- We will continue to improve the compatibility score even if the challenge is finished
- We have 95% compatibility score for TypeScript
- We are currently working on supporting CSS, and we plan to support in a near future HTML, Markdown, Vue
  Any contribution is welcomed :)

Author: > Maybe we could also say that:
> 
>     * We consider any divergence as a bug, and any tester should report divergence
> 

Done

>     * We will continue to improve the compatibility score even if the challenge is finished
> 

How should we phrase it?

>     * We have 95% compatibility score for TypeScript
> 

I am not sure where to put it and how to phrase it. Can you suggest something?

>     * We are currently working on supporting CSS, and we plan to support in a near future HTML, Markdown, Vue

I would prefer to wait on an official roadmap, before promising something that we can't fulfil. 

>       Any contribution is welcomed :)

Where should we have this phrase? How should we phrase it?

Reviewer: > How should we phrase it?
> Where should we have this phrase? How should we phrase it?

Something like:

`
Although the challenge is over, we are committed to improve even more the compatibility score with prettier. Any contribution in this regard is very welcome.
`

What do you think?

```shell
biome --config-path=../path/where/config/is lsp-proxy
biome --config-path=../path/where/config/is start
````

The CLI now exposes the option `--diagnostic-level`, that allows to filter the kind of diagnostics printed to terminal.

```shell
biome check --diagnostic-level=error ./src
```

## New lint rules, and promoted rule

## Homage to our maintainers

Since Biome was forked, new people joined the project. They have been helping with in so many ways that you can't even imagine: new features, side projects, engaging with the community, support, documentation, and more. OSS is not just about coding.

Thank you to:
- [Madeline Gurriarán @SuperchupuDev](https://github.com/SuperchupuDev)
- [Nicolas Hedger @nhedger](https://github.com/nhedger)
- [Victor Teles @victor-teles](https://github.com/victor-teles)

## New sponsors

Last but not least, we are proud to announce that we have two new sponsors"

- Gold: Shiguredō (https://shiguredo.jp/)
- Bronze: KANAME (https://www.kanamekey.com/)

> 

Reviewer: Maybe we could just say that Biome is also a linter with X stable rules with Y recommended rules?

Author: Let's wait for the promoted rules, then you can add something there

---
title: Biome formatter wins the Prettier challenge
template: splash
url: /blog/biome-wins-prettier-challenge
description: Biome formatter now 95% compatible with Prettier
summary: Biome formatter now 95% compatible with Prettier
authorName: Emanuele Stoppa
authorAvatar: https://avatars.githubusercontent.com/u/602478?v=4
authorUrl: https://twitter.com/ematipico
coauthorName: Biome Core Team, Biome Maintainers
coauthorAvatar: /img/circle-indent-logo.svg
pubDate: 2023-11-27
layout: ../../../layouts/Blog.astro
coverImage: blog/prettier-challenge.png
coverImageAltText: The Prettier challenge banner, with the Biome logo over it
socialImage: blog/prettier-challenge.png
---

With the release of Biome **`v1.4.0`**, we claim the bounty of the [Prettier challenge](https://console.algora.io/challenges/prettier)!

With `v1.4.0`, you'll get a better formatter experience, more formatting options, new VSCode features, new sponsors and more!

You can upgrade Biome by running the following command:

```bash
npm install --save-dev --save-exact @biomejs/biome@1.4.0
pnpm update --save-exact @biomejs/biome@1.4.0
yarn upgrade --exact @biomejs/biome@1.4.0
```

##  Better formatter

> Biome formatter has now **over 96% in terms of compatibility** against [Prettier](https://prettier.io/)!

Reviewer: > > Any contribution is welcomed :)
>
> Where should we have this phrase? How should we phrase it?

Maybe at the start? We could just say 95% for these 3 languages. The displayed score on the challenge page corresponds to the global score and thus includes these 3 languages.

```suggestion
Biome formatter has now **over 95% in terms of compatibility** against [Prettier](https://prettier.io/)!
This score is computed for JavaScript, TypeScript, and JSX formatting.
```
